src/shm/pool/slot.rs

```diff
@@ -0,0 +1,501 @@
+//! A pool implementation based on buffer slots
+
+use std::io;
+use std::sync::{
+    atomic::{AtomicU8, AtomicUsize, Ordering},
+    Arc, Mutex, Weak,
+};
+
+use wayland_client::{
+    backend::InvalidId,
+    protocol::{wl_buffer, wl_shm, wl_surface},
+    Proxy,
+};
+
+use super::raw::RawPool;
+
+#[derive(Debug, thiserror::Error)]
+pub enum CreateBufferError {
+    /// Protocol error.
+    #[error(transparent)]
+    Protocol(#[from] InvalidId),
+
+    /// Slot creation error.
+    #[error(transparent)]
+    Io(#[from] io::Error),
+
+    /// Pool mismatch.
+    #[error("Incorrect pool for slot")]
+    PoolMismatch,
+
+    /// Slot size mismatch
+    #[error("Requested buffer size is too large for slot")]
+    SlotTooSmall,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ActivateSlotError {
+    /// Buffer was already active
+    #[error("Buffer was already active")]
+    AlreadyActive,
+
+    /// Protocol error.
+    #[error(transparent)]
+    Protocol(#[from] InvalidId),
+}
+
+#[derive(Debug)]
+pub struct SlotPool {
+    pub(crate) inner: RawPool,
+    free_list: Arc<Mutex<Vec<FreelistEntry>>>,
+}
+
+#[derive(Debug)]
+struct FreelistEntry {
+    offset: usize,
+    len: usize,
+}
+
+/// A chunk of memory allocated from a [SlotPool]
+///
+/// Retaining this object is only required if you wish to resize or change the buffer's format
+/// without changing the contents of the backing memory.
+#[derive(Debug, Clone)]
+pub struct Slot {
+    inner: Arc<SlotInner>,
+}
+
+#[derive(Debug)]
+struct SlotInner {
+    free_list: Weak<Mutex<Vec<FreelistEntry>>>,
+    offset: usize,
+    len: usize,
+    active_buffers: AtomicUsize,
+}
+
+/// A [`wl_buffer::WlBuffer`] allocated from a [SlotPool].
+///
+/// When this object is dropped, the buffer will be destroyed immediately if it is not active, or
+/// upon the server's release if it is.
+#[derive(Debug)]
+pub struct Buffer {
+    buffer: wl_buffer::WlBuffer,
+    height: i32,
+    stride: i32,
```

Comment:
width could be useful if you need to tell a software renderer the extent of the framebuffer.

Reply:
It's easy to track this yourself if it's not a trivial calculation from the stride (most commonly - ARGB8888 with align 1 or 4 - it will be stride/4). The values that are stored are just the ones needed to produce the correct buffer size.

Reply:
If format of the buffer is exposed you could determine the size but otherwise the user has to use a fixed format for all the buffers to know the amount of bytes per pixel. Most of the time I'll be 4 but sometimes it might not be. I don't know how much you value that but it'd be better if the pool was capable of supporting all formats.

Reply:
Tracking formats by storing the used format alongside the buffer seems reasonable to me.  If you support anything beyond {A,X}RGB8888 you already need to track support prior to creating the buffer, so you either store this alongside the pool or you can store it alongside the buffer.

Alternatively the `Buffer` struct could retain the full fields as configured, if this type of tracking is going to be normal for the SCTK front-ends for various protocol objects: having surfaces or windows track their size, scale, and transform, and other state.